vis2k/Mirror
# feat: Network Scene Checker Component

The Network Scene Checker component controls the visibility of game objects for network clients, based on which scene they're in.  This is useful for isolating players and networked objects by subscene, especially in games where the individual subscenes are small, and for multi-match play.

Similar to the Network Proximity Checker, a `ForceHidden` bool is included.

Documentation included.

Test Package
[SceneCheckerTest.zip](https://github.com/vis2k/Mirror/files/4111828/SceneCheckerTest.zip)

![image](https://user-images.githubusercontent.com/9826063/73119130-7d817080-3f2b-11ea-8237-7ed113d5f13f.png)

To run the test, follow these steps carefully:
- New project, import Mirror master, then this PR.
- Restart Unity
- Unzip and import the attached SceneCheckTest Unity pacakge
- **Do not open any scenes yet**
- In Build Settings, remove any other scenes, then add all 3 scenes from the Scenes folder
- Now you can open the Main scene, and verify that Scene1 and Scene2 are in the Subscenes list of the Network Manager
- Play as LAN Host, noting that both subscenes are additively loaded and the player is moved to Scene1 in the hierarchy
- Click the button at the top left of the screen to change subscenes
- Observe that the cube from Scene1 and the sphere of Scene2 are only shown when the player is in their respective subscene.
- Stop and build and run
- Play LAN Host or LAN Client in the editor, the other in the built app
- Click the button at the top left of the screen to change subscenes as before on host and client
- Observe the same as before, plus the other player is shown appropriately in the hierarchy

For this test, player objects are not physically moved to subscenes on the client, only on the server.  It's too easy to lose a player object on a client if it's in a subscene that gets unloaded.  It's only important for the Scene Checker component that the objects are in the correct subscene on the server.
